### ISSUE TITLE
CoreWebView2.AddHostObjectToScript throws System.Exception

### ISSUE THREAD

**Issue by Billeclipse** [MicrosoftEdge/WebView2Feedback#2529]
**Description**
CoreWebView2.AddHostObjectToScript throws System.Exception: 'The group or resource is not in the correct state to perform the requested operation.'

**Version**
SDK: 1.0.1245.22 latest stable - (and Microsoft.UI.Xaml 2.8.0-prerelease.210927001)
Runtime: 
Framework: UWP
OS: Windows 11, build 22000

**Repro Steps**
1. Create an UWP project
2. Add necessary nuget packages (Microsoft.Web.Webview2 latest stable release and Microsoft.UI.Xaml 2.8.0-prerelease)
3. Add WebView2 component to MainPage.xaml
4. Create HostObject class with necessary tags
5. Add code for navigation starting or navigation completed method in WebView2 (MainPage.xaml.cs)
6. Ensure CoreWebView2 is not null by calling myWebView.EnsureCoreWebView2Async(); inside method
7. Try use myWebView.CoreWebView2.AddHostObjectToScript("HostObject", hostObject); inside method

Example of HostObject class:
```
HostObject class example:
[AllowForWeb]
[ClassInterface(ClassInterfaceType.AutoDual)]
[ComVisible(true)]
public class JsCallbacksUwp
{
    public JsCallbacksUwp()
    {

    }
}
```

Example code for Navigation Completed that throws the exception:
```
private async void WebView2_NavigationCompleted(WebView2 sender, CoreWebView2NavigationCompletedEventArgs args)
  {
      await sender.EnsureCoreWebView2Async();

      var browserHostObject = new JSCallbacksUwp();

      sender.CoreWebView2.AddHostObjectToScript("HostObject", browserHostObject);
  }
```
Exception Thrown:

![image](https://user-images.githubusercontent.com/17092072/174022492-39f1c5e9-309e-45e8-af5e-763b5c36cca3.png)

**Additional context**

**Comment by Syul968** [comment]
Hi @Billeclipse,

Could you provide the stack throwing this exception? Also, is there any log in debug output when this happens? This exception could be thrown as a result of the passed object not implementing `IDispatch`. The WinRT `CoreWebView2` class (the one used by WinUI) tries to automatically create a "wrapper" when this is the case, but it needs a `DispatchAdapter` instance to be set through `CoreWebView2Settings.HostObjectDispatchAdapter`.

We have also been working on this feature lately. Our latest pre-release package ([1.0.1248-prerelease](https://www.nuget.org/packages/Microsoft.Web.WebView2/1.0.1248-prerelease)) includes a tool, wv2winrt, to help project new and existing WinRT components: [Call native-side WinRT code from web-side code](https://docs.microsoft.com/en-us/microsoft-edge/webview2/how-to/winrt-from-js?tabs=csharp)

Thanks,
Luis

**Comment by Syul968** [comment]
StackOverflow: [CoreWebView2.AddHostObjectToScript throws System.Exception](https://stackoverflow.com/questions/72683537/corewebview2-addhostobjecttoscript-throws-system-exception)

**Comment by Billeclipse** [comment]
Hey @Syul968, 
Thank you for your answer, I followed the steps at [Call native-side WinRT code from web-side code](https://docs.microsoft.com/en-us/microsoft-edge/webview2/how-to/winrt-from-js?tabs=csharp) and **successfully** called the methods from WebView DevTools at last step although WinRTAdapter was showing red in the InitializeWebView2Async method.

![image](https://user-images.githubusercontent.com/17092072/174981387-4f63944c-d63e-400a-bdd0-75336f3011dc.png)

But isn't there a way `AddHostObjectToScript` with C# WinRT to call my methods through i c# class like the `JsCallbacksUwp` class above?
If not is there any example to create a similar class in the C++ WinRT?
What i am trying to do in my project is to inject js code with `AddScriptToExecuteOnDocumentCreatedAsync` method and then when the user press a button/link inside the WebView that js function that i injected earlier call through my C# HostObject the method that i need in C#. Is there a way to do this through the C++ WinRT? 

Thanks,
Bill

**Comment by david-risney** [comment]
You can set that up although not with a C# class but a C# based WinRT class.

Currently the WinRT API `CoreWebView2.AddHostObjectToScript` requires you to use a WinRT class (that has gone through  wv2winrt as @Syul968 referenced above). The .NET API `CoreWebView2.AddHostObjectToScript` requires a .NET class that supports IDispatch. The COM API `ICoreWebView2::AddHostObjectToScript` requires an IDispatch object. 

You are using XAML which is based in WinRT and so you have to use the WinRT WebView2 APIs and so must use a WinRT class. The WinRT type system doesn't have a way to reference non-WinRT types, so we really need a WinRT type, but it can be a C# based WinRT type. You can use [c++/winrt](https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/) or [c#/winrt](https://docs.microsoft.com/en-us/windows/apps/develop/platform/csharp-winrt/)(or the older [C# built-in WinRT support](https://docs.microsoft.com/en-us/windows/uwp/winrt-components/creating-windows-runtime-components-in-csharp-and-visual-basic)) or really anything that produces WinRT classes (like [rust/winrt](https://github.com/microsoft/windows-rs)) to implement a WinRT class. Then follow the directions @Syul968 referenced to run it through wv2winrt and use it in your project. 

Our current documentation only shows you how to do it for OS provided WinRT APIs and not for your own WinRT APIs. We need to fix this. Until then I would recommend following the existing directions from [Call native-side WinRT code from web-side code](https://docs.microsoft.com/en-us/microsoft-edge/webview2/how-to/winrt-from-js?tabs=csharp), and then additionally doing the following:
1. Add a third project to your VS solution that implements your winrt class.
2. Have the WinRTAdapter project 'Add a reference' to your new third project containing your winrt class.
3. Update the WinRTAdapter project's [`Include filter`](https://docs.microsoft.com/en-us/microsoft-edge/webview2/how-to/winrt-from-js?tabs=csharp#step-6-generate-source-code-for-selected-host-apis) to also include your new class.
4. Add an additional line to `InitializeWebView2Async` to add your winrt class's namespace `WebView2.CoreWebView2.AddHostObjectToScript("MyCustomNamespace", dispatchAdapter.WrapNamedObject("MyCustomNamespace", dispatchAdapter));`
5. And optionally add your namespace sync proxy as a global object in script `window.MyCustomNamespace = chrome.webview.hostObjects.sync.MyCustomNamespace;`

You should then be able to use your type via script.

**Comment by Billeclipse** [comment]
Thank you for your asnwer @david-risney
I set up my C# WinRT class and followed the additional steps from your answer above.
All went through nicely except that it doesn't recognize my object from C# WinRT class inside the javascript injection.
Specifically i added the reference to my new project
![image](https://user-images.githubusercontent.com/17092072/175275643-9b2dc4a0-853b-4bf5-a90f-23fa80ccf604.png)
![image](https://user-images.githubusercontent.com/17092072/175275800-bca5ed1c-ef24-4f4b-a51b-45d7523551ec.png)
and i included the filter in WinRTAdapter
![image](https://user-images.githubusercontent.com/17092072/175275951-2642cf5f-5329-4f84-bd50-1dd2d0952b6e.png)

My Bridge class looks like this:
```
namespace WebViewCommunication
{
    [ComVisible(true)]
    public sealed class Bridge
    {
        public Bridge()
        {
        }
    }
}
```

also i added the following lines to `InitializeWebView2Async` : 
```
private async void InitializeWebView2Async()
 {
   await BookBrowser.EnsureCoreWebView2Async();
   var dispatchAdapter = new WinRTAdapter.DispatchAdapter();
   BookBrowser.CoreWebView2.AddHostObjectToScript("Windows", dispatchAdapter.WrapNamedObject("Windows", dispatchAdapter));
   BookBrowser.CoreWebView2.AddHostObjectToScript("WebViewCommunication.Bridge", dispatchAdapter.WrapNamedObject("WebViewCommunication.Bridge", dispatchAdapter));
    await BookBrowser.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync(AppConstants.INJECTION);
 }
```
AppConstants.INJECTION is my js code which if add the following line:
`window.WebViewCommunication.Bridge = chrome.webview.hostObjects.sync.WebViewCommunication.Bridge;`

It throws error in console: 
![image](https://user-images.githubusercontent.com/17092072/175274545-fb0fe0fb-09d9-414f-826e-999e90d525dc.png)

Also tried to remove the line above but of course then brings the error below:
`if (window.WebViewCommunication.Bridge)`

![image](https://user-images.githubusercontent.com/17092072/175272106-bc03757c-d61f-4ed7-80ab-2f49772cd360.png)

Am i doing something wrong with my C# WinRT class?

**Comment by Billeclipse** [comment]
Thank you for your answer @david-risney.
I set up my C# WinRT class and followed the additional steps from your answer above.
All went through nicely and although i had some problems with javascript code after the changes i looked up the original guide for [Call native-side WinRT code from web-side code](https://docs.microsoft.com/en-us/microsoft-edge/webview2/how-to/winrt-from-js?tabs=csharp) and solved my last problem too. So a big thanks again to @Syul968 's answer.
Below i will show what i did at my project after @david-risney steps just for future reference to anyone having the same issue.

1. I created my new C# WinRT project and my new C# WinRT class

C# WinRT Project:
![image](https://user-images.githubusercontent.com/17092072/175275643-9b2dc4a0-853b-4bf5-a90f-23fa80ccf604.png)

C# WinRT Class:
```
namespace WebViewCommunication
{
    [ComVisible(true)]
    public sealed class Bridge
    {
        public Bridge()
        {
        }
       
       public void RetrieveScore(int score)
       {
            Debug.WriteLine("Score: " + score);
       }
    }
}
```

2. I added the reference to my new project

![image](https://user-images.githubusercontent.com/17092072/175275800-bca5ed1c-ef24-4f4b-a51b-45d7523551ec.png)

3. i added the following lines to `InitializeWebView2Async` : 

```
private async void InitializeWebView2Async()
 {
   await BookBrowser.EnsureCoreWebView2Async();
   var dispatchAdapter = new WinRTAdapter.DispatchAdapter();
   BookBrowser.CoreWebView2.AddHostObjectToScript("Windows", dispatchAdapter.WrapNamedObject("Windows", dispatchAdapter));
   BookBrowser.CoreWebView2.AddHostObjectToScript("WebViewCommunication.Bridge", dispatchAdapter.WrapNamedObject("WebViewCommunication.Bridge", dispatchAdapter));
   await BookBrowser.CoreWebView2.AddScriptToExecuteOnDocumentCreatedAsync(AppConstants.INJECTION);
 }
```

4. `AppConstants.INJECTION` is my js code in which i added the following lines (example just to call a method with parameter from my bridge object when a mouse click is pressed):
```
const Bridge = chrome.webview.hostObjects.sync.WebViewCommunication.Bridge;
document.addEventListener("click", function (e) {
   if (Bridge) 
   {
       (new Bridge).retrieveScore(25);
   }
}
```
Hopefully in next patches this problem can be solved in a easier and more clean way but for now this solve will do.

**Comment by david-risney** [comment]
Yes, great question that we need to document =).

After going through [the steps in our docs](https://docs.microsoft.com/en-us/microsoft-edge/webview2/how-to/winrt-from-js?tabs=csharp), you should be using our synchronous proxies. For async method calls, there's a property mentioned in those steps `chrome.webview.hostObjects.options.forceAsyncMethodMatches`. This is an array of regex that if any match a method name on a sync proxy, the method will be run asynchronously instead. Setting this to `[/Async$/]` will have it match any method ending with the suffix Async. Then matching method calls work just like a method on an async proxy and returns a promise that you can await.

Something like the following (the file picker part isn't real working code but the async method call should match)

```
const Windows = chrome.webview.hostObjects.sync.Windows;
chrome.webview.hostObjects.options.forceAsyncMethodMatches = [/Async$/];

let fileOpenPicker = new Windows.Storage.Pickers.FileOpenPicker();
let result = await fileOpenPicker.pickSingleFileAsync();
```

**Comment by david-risney** [comment]
Sorry, I was trying to show what the syntax for an async method could look like. I don't know if the file picker APIs specifically would work in your case. The above should be the same as if you had run that same code on the UI thread (the same one that the WebView2 is on) at that time. I believe the file picker has some focus requirements and maybe other requirements. If you run that same code on your UI thread without WebView2 involved does it work?

**Comment by zh4ui** [comment]
Thanks, you're right. The problem is on FileOpenPicker. Investigation shows it's a problem of FileOpenPicker itself. In WinUI3 it has to be initialized with a HWND:

```cpp
        auto picker = Windows::Storage::Pickers::FileOpenPicker{};
        winrt::check_hresult(picker.as<IInitializeWithWindow>()->Initialize(GetActiveWindow()));
        picker.FileTypeFilter().Append(L"*");
        co_await picker.PickSingleFileAsync();
```


I was wondering that must wv2winrt be used in a  Windows Runtime Component project that generates DLL? Can it be a static lib, or co-exist with the main project?  As far as I can remember, this is not possible in the original Edge WebView.

Anyway I tried with a static lib and it seemed to work.

**Comment by DDHSchmidt** [comment]
Hello @david-risney 

> [...] or [**c#/winrt**](https://docs.microsoft.com/en-us/windows/apps/develop/platform/csharp-winrt/)[...] Then follow the directions @Syul968 referenced to run it through wv2winrt and use it in your project.
> 
> Our current documentation only shows you how to do it for OS provided WinRT APIs and not for your own WinRT APIs. We need to fix this. Until then I would recommend following the existing directions from [Call native-side WinRT code from web-side code](https://docs.microsoft.com/en-us/microsoft-edge/webview2/how-to/winrt-from-js?tabs=csharp), and then additionally doing the following:
> 
> [...]
> 2. Have the WinRTAdapter project 'Add a reference' to your new third project containing your winrt class.
> **3. Update the WinRTAdapter project's [`Include filter`](https://docs.microsoft.com/en-us/microsoft-edge/webview2/how-to/winrt-from-js?tabs=csharp#step-6-generate-source-code-for-selected-host-apis) to also include your new class**.
> **4. Add an additional line to `InitializeWebView2Async` to add your winrt class's namespace `WebView2.CoreWebView2.AddHostObjectToScript("MyCustomNamespace", dispatchAdapter.WrapNamedObject("MyCustomNamespace", dispatchAdapter));`**
> [...]

I'm currently trying to upgrade our Xamarin.Forms UWP target to WebView2 (Please don't suggest a migration to MAUI. That will eventually come, but much later and is another beast to tackle...) and boy does this keep my on my toes.
The paragraphs I rendered bold baffle me, since when using a c#/winrt project template there is no property dialog anymore that would allow "Include filters" to be set and no mention of wv2winrt anymore?

Basically I want to use a c# class (with the attributes ClassInterface(ClassInterfaceType.AutoDual)
    and ComVisible(true) ) from a .Net Standard 2.0 project as the host object for my WebView2 but whenever I try I get the aforementioned System.Exception: 'The group or resource is not in the correct state to perform the requested operation.'

Going the c++/winrt --> wv2winrt route doesn't work since I can't reference my c# class in the c++ project and I described my woes with the c#/winrt approach a pargraph earlier.

Is my use-case too exotic? Did I miss something?

**Comment by koenvd** [comment]
I also have issues to get this working @david-risney @Syul968 but I'm also a bit confused tbh.

So my app is a WinUI3.0 **Win32** desktop app (no UWP) and my goal is to call the WinRT API from Webview2.  

So I still went through the flow as described above using the wv2winrt tool. (Here I'm a bit confused if this is the way to go as well for a Win32 desktop app.)

I was able to reference the `WinRTAdapter.DispatchAdapter` but I always end up with the error "The type or namespace name 'WinRTAdapter' could not be found (are you missing a using directive or an assembly reference?)" while the referenced project file shows up in the solution without any errors. A bit similar as @Billeclipse screenshot. I also needed to add a reference to the CSWinRT package to resolve an error of referencing a winmd file when building for .net version higher then 5.

Also changed the code to call the interop version of `AddHostObjectToScript` as mentioned in a linked issue.

![Screenshot 2022-08-22 184302](https://user-images.githubusercontent.com/15209987/185974505-6008e1d7-a99a-4b76-83dd-684be39c58db.png)

I also tried to add the Webview2 prerelease package to my class library but this didn't resolve the issue. I then also receive more error messages because the WindowsAppSDK and the Webview2 prerelease package are exposing the same Webview2 interface.

So just wondering if what I'm doing is the correct flow to get access to the Windows API in a WinUI3.0 desktop app? 

If it is also wondering if you might have an idea why the namespace cannot be found? And if not I probably need to take the other route (https://docs.microsoft.com/en-us/microsoft-edge/webview2/how-to/hostobject?tabs=dotnetcsharp) to get access to the Windows API but then as well wondering if there's a similar tool as well to generate the 'bridge' code for the Windows APIs I want to use.

**Comment by MichaelLiddiard** [comment]
Im seeing the same issues as @koenvd in a fresh Win UI 3 project. Is there any specific guidance for a Win32 not UWP host?

**Comment by nsandeep87** [comment]
Even I am facing the same issue that @koenvd has raised. 
@david-risney :   We are trying to migrate UWP+JS project to WinUI3 and Webview2. Is it safe to assume  that WinUI3 can consume wv2winrt, but but at this point due to some glitch we are not able to find the namespace. Also could you please  provide any guidance as to why the namespace is not available.

**Comment by Diegorro98** [comment]
Hi @david-risney 
Any news on the WinUI3 sample and docs? I face to the same blocking problem than @koenvd  and @nsandeep87

**Comment by Take-A-Byte** [comment]
Hi @koenvd , I am facing the same problem as you did back in August. Did following [the other route](https://docs.microsoft.com/en-us/microsoft-edge/webview2/how-to/hostobject?tabs=dotnetcsharp) work for you? How was the experience with it?
If it did not work, did you find any solution that worked for you? I would appreciate it if you could share it with us.

By the way, @david-risney @Syul968 - could you please share approximately by when can we expect an update on this? I think everyone here would highly appreciate it if you could share a timeline...

**Comment by Take-A-Byte** [comment]
Created a new issue https://github.com/MicrosoftEdge/WebView2Feedback/issues/2937 as this one was closed

**Comment by Diegorro98** [comment]
Great idea

**Comment by Take-A-Byte** [comment]
FYI: https://github.com/microsoft/microsoft-ui-xaml/issues/7170#issuecomment-1215910370 example in this comment solves the above problem.
